### DISCUSSION TITLE
I want to lazy load Adsense

### DISCUSSION THREAD

**Discussion by kuroroblog** [gatsbyjs/gatsby#32835] [Help]
I run the following sites and have google adsense installed.

site : https://kuroro.blog/

I am using the `gatsby-plugin-google-adsense` module to introduce google adsense.

However, because the page speed insight score is bad, I thought about lazyloading against google adsense.

<img width="1680" alt="screenshot 2021-08-17 21 59 49" src="https://user-images.githubusercontent.com/23373288/129730027-090493a5-4948-4389-959e-c3f3bda5ad42.png">

I thought as follows as an implementation method.

1. Get rid of `gatsby-plugin-google-adsense` plugin in `gatsby-config.js`.
2. Write javascript to lazy load adsense.

Reference example of 2

``` js
<script
     dangerouslySetInnerHTML={{
       __html: `
          (function(window, document) {
            function main() {
              // GoogleAdSense読込み
              var ad = document.createElement('script');
              ad.type = 'text/javascript';
              ad.async = true;
              // 新コードの場合、サイト運営者IDを書き換えてコメントアウトを外す
              ad.dataset.adClient = 'ca-pub-xxxxxxxxxxxxxxxx';
              ad.src = 'https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js';
              var sc = document.getElementsByTagName('script')[0];
              sc.parentNode.insertBefore(ad, sc);
            }
        
            // 遅延読込み
            var lazyLoad = false;
            function onLazyLoad() {
              if (lazyLoad === false) {
                // 複数呼び出し回避 + イベント解除
                lazyLoad = true;
                window.removeEventListener('scroll', onLazyLoad);
                window.removeEventListener('mousemove', onLazyLoad);
                window.removeEventListener('mousedown', onLazyLoad);
                window.removeEventListener('touchstart', onLazyLoad);
                window.removeEventListener('keydown', onLazyLoad);
                main();
              }
            }
            window.addEventListener('scroll', onLazyLoad);
            window.addEventListener('mousemove', onLazyLoad);
            window.addEventListener('mousedown', onLazyLoad);
            window.addEventListener('touchstart', onLazyLoad);
            window.addEventListener('keydown', onLazyLoad);
            window.addEventListener('load', function() {
              // ドキュメント途中（更新時 or ページ内リンク）
              if (window.pageYOffset) {
                onLazyLoad();
              }
              //何もアクションがないときは指定秒数後に読み込み開始（ミリ秒）
              window.setTimeout(onLazyLoad,3000)
            });
          })(window, document);
        `
     }}
/>
```

2 references
- https://note.com/karaage1970/n/n5f52c8ffea5b
- https://gan0803.dev/blog/2021-01-06-lazy-load-adsense
- https://www.bugbugnow.net/2019/05/GoogleAdSense-lazy-loading.html

As a result of implementation, the script created in 2 was inserted, but the script for adsense with `data-checked-head="true"` was automatically added.

``` js
<script data-ad-client="ca-pub-***************" async="" src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js" data-checked-head="true"></script>
```

I think that additional attributes were added at render time, probably because Google couldn't authenticate the site.

Reference
- https://support.google.com/adsense/thread/85862251/adsense%E7%94%B3%E8%AB%8B%E6%99%82%E3%81%AB%E3%80%81%E3%83%98%E3%83%83%E3%83%80%E3%83%BC%E3%81%AB%E8%A8%98%E8%BC%89%E3%81%99%E3%82%8B%E3%83%AA%E3%83%B3%E3%82%AF%E3%81%AB%E8%87%AA%E5%8B%95%E3%81%A7data-checked-head-true-%E3%81%8C%E3%81%A4%E3%81%84%E3%81%A6%E3%81%97%E3%81%BE%E3%81%86?hl=ja

# Discussion
- How to lazy load adsense?(I'm glad to have a module)
- I would like to talk about what measures are taken for 3rd party files in page speed insight.

**Comment by krismorf** [comment]
@kuroroblog you can use `gatsby-plugin-google-tagmanager`. In Tag Manager you can set a [Timer Trigger](https://support.google.com/tagmanager/answer/7679323?hl=en) to fire your Ad Sense Tag.

**Reply by kuroroblog** [reply]
> gatsby-plugin-google-tagmanager
I see, Try it.

Thanks!! @krismorf
